### ISSUE TITLE
injective or bijective?

### ISSUE THREAD

**Issue by yuffon** [jhjacobsen/pytorch-i-revnet#17]
In the paper, there are two networks, injective and bijective.
It seems that this repo only includes the bijective network.
Is that right?

**Comment by jhjacobsen** [comment]
The script to train the injective ImageNet network is not part of the repo, but you can easily write it yourself. The Cifar10 example is an injective network, which you can use as an example.

**Comment by yuffon** [comment]
> The script to train the injective ImageNet network is not part of the repo, but you can easily write it yourself. The Cifar10 example is an injective network, which you can use as an example.

The cifar10 example and imagenet example use only one different init_ds argument. It seems that the cifar10 example is also a bijective network.

**Comment by jhjacobsen** [comment]
The Cifar10 network uses what we call "injective padding", so channel size is increased by padding zeros in intermediate layers. This is automatically done in the code whenever you choose the channel size of the revnet block to be larger than the channel size of the input. So you can easily build injective nets with the current codebase.

**Comment by yuffon** [comment]
> The Cifar10 network uses what we call "injective padding", so channel size is increased by padding zeros in intermediate layers. This is automatically done in the code whenever you choose the channel size of the revnet block to be larger than the channel size of the input. So you can easily build injective nets with the current codebase.

Do you mean that, if I set the nChannels as [3, 12, 48], the network would be bijective for cifar10?
If this holds, the network would be too narrow to keep the universal approximating ability.

**Comment by jhjacobsen** [comment]
The example:

`python CIFAR_main.py --nBlocks 18 18 18 --nStrides 1 2 2 --nChannels 16 64 256`

Is giving you an injective network, the 3 channel Cifar10 input is padded with 13 channels consisting of zeros.

**Comment by yuffon** [comment]
Thank you for you reply. I get your point. The first number in the nChannels arguments determines whether the network is injective or bijective, is that right? If yes, the two commands for cifar10 and imagenet both use injective padding at the first layer. So they are both injective networks. What if I want to use a totally bijective network? I set nChannels as 3, 12,48, it reports errors.

**Comment by jhjacobsen** [comment]
The example:

`python ILSVRC_main.py --data /path/to/ILSVRC2012/ --nBlocks 6 16 72 6 --nStrides 2 2 2 2 --nChannels 24 96 384 1536 --init_ds 2`

Is bijective. init_ds: 3 channel input -> 12 channel input, nStrides 2 in first stage: 12 channel input -> 48 channel input. Split dimensions: 48 channel input -> 2 * 24 channel input. Hence, input channels of iRevNet block are 24.

**Comment by yuffon** [comment]
I see. Thanks a lot. The init downsampling implemented in psi() is important. It increases the initial channels. The imagenet resolution is large enough so that mulitple downsampling ops can be used. Have you tried the bijective network on cifar10?

**Comment by jhjacobsen** [comment]
Exactly, you're welcome.

**Comment by yuffon** [comment]
I have tried the bijective network config (originally for imagenet) on cifar10. The argument is --nBlocks 6 16 72 6 --nStrides 2 2 2 2 --nChannels 24 96 384 1536 --init_ds 2. It only reaches 82% test accuracy and 91% train accuracy. The network is larger than the injective one used for cifar10. I don't know why.

**Comment by jhjacobsen** [comment]
This is roughly what I found as well. Reasons for this are quite deep and will be answered in future research.

**Comment by yuffon** [comment]
What I need is a bijective network good enough for cifar10. Have you tried some good bijective network for cifar10? By the way, the psi op reshapes the input without considering the spatial relationship. Have you tried to use the reshape op like figure 3 in realNVP paper?

**Comment by jhjacobsen** [comment]
These are problems unrelated to this repository and open research questions themselves. I will stop answering here, but feel free to follow up via email: j.jacobsen@vectorinstitute.ai

**Comment by shuizidesu** [comment]
The injective padding operation is also invertible, and the class 'injective_pad' also have the 'inverse' method. So I think we can still reconstruct the input images from the final representation. I'm confused about why to call it an injective network. Thanks for your answer.
